client/mysqldump.c

```diff
@@ -4165,6 +4252,397 @@ static char *getTableName(int reset)
+*/
+
+static int dump_create_user(const char *user)
+{
+  DYNAMIC_STRING sqlbuf;
+  MYSQL_ROW row;
+  MYSQL_RES *tableres;
+
+  init_dynamic_string_checked(&sqlbuf, "SHOW CREATE USER ", 256, 1024);
+  dynstr_append_checked(&sqlbuf, user);
+
+  if (mysql_query_with_error_report(mysql, &tableres, sqlbuf.str))
+  {
+    dynstr_free(&sqlbuf);
+    return 1;
+  }
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    fprintf(md_result_file, "CREATE %sUSER %s%s;\n", opt_replace_into ? "/*M!100103 OR REPLACE */ ": "",
+            opt_ignore ? "IF NOT EXISTS " : "",
+            row[0] + sizeof("CREATE USER"));
+  }
+  mysql_free_result(tableres);
+  dynstr_free(&sqlbuf);
+  return 0;
+}
+
+
+
+/*
+  dump all users and roles
+*/
+
+static int dump_all_users()
+{
+  MYSQL_ROW row;
+  MYSQL_RES *tableres;
+  int result= 0;
+  /* Roles added in MariaDB-10.0.5 or MySQL-8.0 */
+  my_bool maria_roles_exist= (mysql_get_server_version(mysql) >= 100005);
+  my_bool mysql_roles_exist= (mysql_get_server_version(mysql) >= 80001) && !maria_roles_exist;
+
+  if (mysql_query_with_error_report(mysql, &tableres,
+                                    "SELECT CONCAT(QUOTE(u.user), '@', QUOTE(u.Host)) AS u "
+                                    "FROM mysql.user u "
+				    " /*!80001 LEFT JOIN mysql.role_edges e ON u.user=e.from_user AND u.host=e.from_host "
+				    "         WHERE e.from_user IS NULL */"
+				    " /*M!100005 WHERE is_role='N' */"))
+    return 1;
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    if (opt_replace_into)
+      /* Protection against removing the current import user */
+      /* MySQL-8.0 export capability */
+      fprintf(md_result_file,
+	"DELIMITER |\n"
+        "/*M!100101 IF current_user()=\"%s\" THEN\n"
+        "  SIGNAL SQLSTATE '45000' SET MYSQL_ERRNO=30001, MESSAGE_TEXT=\"Don't remove current user %s'\";\n"
+        "END IF */|\n"
+	"DELIMITER ;\n"
+        "/*!50701 DROP USER IF EXISTS %s */;\n", row[0], row[0], row[0]);
+    if (dump_create_user(row[0]))
+      result= 1;
+    /* if roles exist, defer dumping grants until after roles created */
+    if (!(maria_roles_exist || mysql_roles_exist) && dump_grants(row[0]))
+      result= 1;
+  }
+  mysql_free_result(tableres);
+
+  if (!(maria_roles_exist || mysql_roles_exist))
+    goto exit;
+
+  /* preserve current role. Create a new role for importing that becomes the
+   * default admin for new roles, so it can be dropped after roles are assigned */
+  fputs("SELECT COALESCE(CURRENT_ROLE(),'NONE') into @current_role;\n"
+        "CREATE ROLE IF NOT EXISTS mariadb_dump_import_role;\n"
+	"GRANT mariadb_dump_import_role TO CURRENT_USER();\n"
+        "SET ROLE mariadb_dump_import_role;\n"
+        , md_result_file);
+  /* No show create role yet, MDEV-22311 */
+  /* Roles, with user admins first, then roles they adminster, and recurse on that */
+  if (maria_roles_exist && mysql_query_with_error_report(mysql, &tableres,
+      "WITH RECURSIVE create_role_order AS"
```

Comment:
Probably here and above to put query into buffer.

Reply:
I'm not sure why. The queries are static.